### PR TITLE
Do bundled read in compaction copy phase.

### PR DESCRIPTION
**Pull request by zzmao**
To avoid multiple small IO read and improve disk utilization.

### CODE REVIEW

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 1]
```diff
@@ -86,6 +91,7 @@ public boolean accept(File dir, String name) {
   private StoreFindToken recoveryStartToken = null;
   private CompactionLog compactionLog;
   private volatile CountDownLatch runningLatch = new CountDownLatch(0);
+  private ByteBuffer bundleReadBuffer;
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: `final`?

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 2]
```diff
@@ -120,6 +126,8 @@ public boolean accept(File dir, String name) {
     this.sessionId = sessionId;
     this.incarnationId = incarnationId;
     fixStateIfRequired();
+    this.bundleReadBuffer =
+        ByteBuffer.allocate(Math.max(config.storeCleanupOperationsBytesPerSec * 2, 10 * 1024 * 1024));
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: should we make this a config?

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 3]
```diff
@@ -120,6 +126,8 @@ public boolean accept(File dir, String name) {
     this.sessionId = sessionId;
     this.incarnationId = incarnationId;
     fixStateIfRequired();
+    this.bundleReadBuffer =
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: minor: don't need `this`

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 4]
```diff
@@ -733,52 +741,83 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+      while (srcIndexEntryIterator.hasNext()) {
+        // reset startOffset to begin a new bundle read.
+        long startOffset = -1;
+        List<IndexEntry> readyEntries = new ArrayList<>();
+        // get entries as many as possible and do a bunch read.
+        while (srcIndexEntryIterator.hasNext()) {
+          IndexEntry currentEntry = srcIndexEntryIterator.next();
+          if (startOffset == -1) {
+            startOffset = currentEntry.getValue().getOffset().getOffset();
+            readyEntries.clear();
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: is this required? Won't it be a new array list already?

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 5]
```diff
@@ -733,52 +741,83 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+                Channels.newChannel(new ByteBufferOutputStream(bufferToUse)));
+            break;
+          }
+        }
+        // copy from buffer to tgtLog
+        for (IndexEntry srcIndexEntry : readyEntries) {
+          IndexValue srcValue = srcIndexEntry.getValue();
+          long usedCapacity = tgtIndex.getLogUsedCapacity();
+          if (isActive && (tgtLog.getCapacityInBytes() - usedCapacity >= srcValue.getSize())) {
+            fileChannel.position(srcValue.getOffset().getOffset());
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: You don't have to position the file channel anymore I think?

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 6]
```diff
@@ -733,52 +741,83 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+            Offset endOffsetOfLastMessage = tgtLog.getEndOffset();
+            // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
+            diskIOScheduler.getSlice(DiskManager.CLEANUP_OPS_JOB_NAME, DiskManager.CLEANUP_OPS_JOB_NAME,
+                writtenLastTime);
+            bufferToUse.position((int) (srcValue.getOffset().getOffset() - startOffset));
+            bufferToUse.limit(bufferToUse.capacity());
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: should it we limit it to the size of the current entry to avoid problems?

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 7]
```diff
@@ -733,52 +741,83 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+            tgtLog.appendFrom(Channels.newChannel(new ByteBufferInputStream(bufferToUse)), srcValue.getSize());
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: if you limit correctly, `Log` has a `appendFrom(ByteBuffer)` method that can be used here (without using channels).

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 1 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: yes.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 2 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: You mean another individual config other than `storeCleanupOperationsBytesPerSec `? I prefer to have a individual config for the bufferSize.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 3 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Got it.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 4 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: yes.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 5 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: yes.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 7 above)
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Sure. Will use it in this way.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 8]
```diff
@@ -733,52 +741,83 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+        while (srcIndexEntryIterator.hasNext()) {
+          IndexEntry currentEntry = srcIndexEntryIterator.next();
+          if (startOffset == -1) {
+            startOffset = currentEntry.getValue().getOffset().getOffset();
+            readyEntries.clear();
+          }
+          readyEntries.add(currentEntry);
+          long totalSize =
+              currentEntry.getValue().getOffset().getOffset() + currentEntry.getValue().getSize() - startOffset;
+          if (totalSize > config.storeCleanupOperationsBytesPerSec || !srcIndexEntryIterator.hasNext()) {
```
Comment (zzmao) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: The logic here is not correct. It allocates byteBuffer most of time.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 2 above)
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: yeah - we still have the `max()` code is what I'm thinking but have a "compaction.min.buffer.size" instead of harcoding to 10.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 9]
```diff
@@ -733,52 +739,95 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+            // only have one in readyEntries and it exceeds bundleReadBuffer's capacity.
+            bufferToUse = ByteBuffer.allocate((int) readySize);
+            fileChannel.transferTo(startOffset, readySize,
+                Channels.newChannel(new ByteBufferOutputStream(bufferToUse)));
+            break;
+          }
+          long nextSize = Integer.MAX_VALUE;
+          if (srcIndexEntryIterator.hasNext()) {
+            // look ahead
+            IndexEntry nextEntry = srcIndexEntryIterator.next();
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: I _think_ there is a bug here. I feel like this logic may be simpler if just iterate over index entries array using an index (or using it as a queue). Let's chat today and see what we can come up with.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 10]
```diff
@@ -120,6 +125,7 @@ public boolean accept(File dir, String name) {
     this.sessionId = sessionId;
     this.incarnationId = incarnationId;
     fixStateIfRequired();
+    bundleReadBuffer = ByteBuffer.allocate(config.storeCleanupOperationsBufferSize);
```
Comment (cgtz) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: I wonder if we should use `ByteBuffer.allocateDirect(...)` in this case since the buffer is long-lived and it could potentially enable some optimizations when used with `FileChannel`

https://stackoverflow.com/questions/5670862/bytebuffer-allocate-vs-bytebuffer-allocatedirect

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 11]
```diff
@@ -733,52 +739,95 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+          if (srcIndexEntryIterator.hasNext()) {
+            // look ahead
+            IndexEntry nextEntry = srcIndexEntryIterator.next();
+            nextSize = nextEntry.getValue().getOffset().getOffset() + nextEntry.getValue().getSize() - startOffset;
+            srcIndexEntryIterator.previous();
+          }
+          if (nextSize > bundleReadBuffer.capacity()) {
+            bufferToUse = bundleReadBuffer;
+            bufferToUse.clear();
+            fileChannel.transferTo(startOffset, readySize,
```
Comment (cgtz) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Should we just use `FileChannel.read()` in this case instead of wrapping the `ByteBuffer` in a channel?

**Comment by zzmao** [comment]
Some tests, for example, `allEntryTypesTest` creates some large record by ``` long lastRecSize = state.log.getSegmentCapacity() - state.index.getCurrentEndOffset().getOffset();```, therefor no extra test was added for this case.

**Comment by vgkholla** [comment]
@zzmao a compaction test has failed in the travis build - can you check?

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java [hunk 12]
```diff
@@ -71,6 +71,8 @@
 
   private MetricRegistry metricRegistry;
 
+  private final ByteBuffer bundleReadBuffer = ByteBuffer.allocateDirect(100);
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java]: I think this should be something like `2 * CuratedLogIndexState.PUT_RECORD_SIZE + 1` to allow for two put entries to exist on the same buffer

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java [hunk 13]
```diff
@@ -1478,14 +1480,15 @@ private void verifyDataPostCompaction(Set<MockId> idsInCompactedLogSegments, lon
    * @param indexEntries the index entries to process.
    * @param logEntriesInOrder the list of {@link LogEntry} instances to add to.
    */
-  private void addToLogEntriesInOrder(List<IndexEntry> indexEntries, List<LogEntry> logEntriesInOrder) {
+  private void addToLogEntriesInOrder(List<IndexEntry> indexEntries, List<LogEntry> logEntriesInOrder,
+      Offset indexSegmentStartOffset) {
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java]: remove the `Offset indexSegmentStartOffset` parameter

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java [hunk 14]
```diff
@@ -2061,7 +2066,7 @@ public boolean equals(Object o) {
       }
 
       LogEntry logEntry = (LogEntry) o;
-      return id.equals(logEntry.id) && entryType == logEntry.entryType;
+      return id.equals(logEntry.id) && entryType == logEntry.entryType && size == logEntry.size;
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java]: also add this to the `hashcode()` method (can be autogenerated by intellij)

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java [hunk 15]
```diff
@@ -2070,5 +2075,10 @@ public int hashCode() {
       result = 31 * result + entryType.hashCode();
       return result;
     }
+
+    @Override
+    public String toString() {
+      return "[id: " + this.id + " entryType: " + this.entryType + " size: " + this.size + "]";
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java]: minor: don't need `this.` qualifiers

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java [hunk 16]
```diff
@@ -340,7 +340,8 @@ public void testClusterManagerWriteStatusDelegateUse() throws StoreException, IO
       //Advance time by 8 days, call compaction to compact segments with deleted data, then verify
       //that the store is now read-write
       time.sleep(TimeUnit.DAYS.toMillis(8));
-      store.compact(store.getCompactionDetails(new CompactAllPolicy(defaultConfig, time)));
+      store.compact(store.getCompactionDetails(new CompactAllPolicy(defaultConfig, time)),
+          ByteBuffer.allocateDirect(100));
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreTest.java]: same here - let's do `2 * PUT_RECORD_SIZE + 1`

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 17]
```diff
@@ -717,6 +723,27 @@ private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws Sto
         IndexValue.Flags.Delete_Index));
   }
 
+  /**
+   * Calculate the farthest index can be used for a bundle IO read based on startIndex and bundleReadBuffer capacity.
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: minor: "Calculate the farthest index _that_ can be used.."

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 18]
```diff
@@ -717,6 +723,27 @@ private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws Sto
...
+   * @param srcIndexEntries all available index entries.
+   * @param startIndex the startIndex for a bundle read.
+   * @return the endIndex which can be used for a bundle read.
+   */
+  private int getBundleReadEndIndex(List<IndexEntry> srcIndexEntries, int startIndex) {
+    long startOffset = srcIndexEntries.get(startIndex).getValue().getOffset().getOffset();
+    int endIndex; // [startIndex, endIndex) for a bundle of read.
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: avoid same line comments

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 19]
```diff
@@ -733,53 +760,89 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+      int startIndex = 0;
+      while (startIndex < srcIndexEntries.size()) {
+        // try to do a bundle of read to reduce disk IO
+        ByteBuffer bufferToUse;
+        long startOffset = srcIndexEntries.get(startIndex).getValue().getOffset().getOffset();
+        int endIndex;
+        if (bundleReadBuffer == null
+            || srcIndexEntries.get(startIndex).getValue().getSize() > bundleReadBuffer.capacity()) {
+          // can't use the preAllocated buffer
+          endIndex = startIndex;
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: maybe a metric (and a trace log message with the id and the size) here will be useful

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 20]
```diff
@@ -733,53 +760,89 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+          bufferToUse.position(0);
+          bufferToUse.limit((int) (
+              srcIndexEntries.get(endIndex).getValue().getOffset().getOffset() + srcIndexEntries.get(endIndex)
+                  .getValue()
+                  .getSize() - startOffset));
+        }
+        // do IO read
+        int bytesRead = fileChannel.read(bufferToUse, startOffset);
+        if (bytesRead != bufferToUse.limit()) {
+          logger.warn("fileChannel is reading more than expected: {}/{}", bufferToUse.limit(), bytesRead);
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: metric also

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 21]
```diff
@@ -733,53 +760,89 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
-          // this is the extra segment, so it is ok to run out of space.
-          logger.info(
-              "There is no more capacity in the destination log in {}. Total capacity is {}. Used capacity is {}."
-                  + " Segment that was being copied is {}", storeId, totalCapacity, usedCapacity,
-              logSegmentToCopy.getName());
-          copiedAll = false;
-          break;
         }
+        if (!copiedAll) {
+          break; // break while loop
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: no same line comments

**Comment by vgkholla** [comment]
Also, there is no test where `bundleReadBuffer` is `null` acc to the code coverage report - will that be easy to add? (Maybe make the `basicTest()` run twice - once with a `null` buffer and once with a non null)

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 22]
```diff
@@ -733,53 +761,93 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
-              "There is no more capacity in the destination log in {}. Total capacity is {}. Used capacity is {}."
-                  + " Segment that was being copied is {}", storeId, totalCapacity, usedCapacity,
-              logSegmentToCopy.getName());
-          copiedAll = false;
+        }
+        if (!copiedAll) {
+          // break outer while loop
           break;
         }
+        startIndex = endIndex + 1;
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: could you add `tgtLog.flush()` just before this line? It might help keep the dirty pages in the buffer cache to a minimum.

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java [hunk 23]
```diff
@@ -2040,15 +2057,17 @@ void dropSegment(String segmentName, boolean decreaseUsedSegmentCount) throws IO
      * The type of the entry.
      */
     PersistentIndex.IndexEntryType entryType;
+    long size;
 
     /**
      * Create an instance with {@code id} and {@code entryType}
      * @param id the {@link MockId} of the entry.
      * @param entryType the type of the entry.
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java]: minor: add javadoc for `size`

**Comment by zzmao** [comment]
@pnarayanan Reminder to review.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 24]
```diff
@@ -733,53 +761,94 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+          bufferToUse = bundleReadBuffer;
+          bufferToUse.position(0);
+          bufferToUse.limit((int) (
+              srcIndexEntries.get(endIndex).getValue().getOffset().getOffset() + srcIndexEntries.get(endIndex)
+                  .getValue()
+                  .getSize() - startOffset));
+        }
+        // do IO read
+        int bytesRead = fileChannel.read(bufferToUse, startOffset);
+        if (bytesRead != bufferToUse.limit()) {
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: What is this for? Are you checking if `bytesRead > bufferToUse.limit()`? That will never happen for two reasons. `ByteBuffer` does not allow a write that goes beyond the `limit` - an attempt to do that will lead to `BufferOverflowException`. Additionally, `FileChannel::read(ByteBuffer dst)` attempts to read at most `dst.remaining()` bytes into `dst`, as per its contract.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 25]
```diff
@@ -717,6 +723,28 @@ private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws Sto
...
+  /**
+   * Calculate the farthest index that can be used for a bundle IO read based on startIndex and bundleReadBuffer capacity.
+   * @param srcIndexEntries all available index entries.
+   * @param startIndex the startIndex for a bundle read.
+   * @return the endIndex which can be used for a bundle read.
+   */
+  private int getBundleReadEndIndex(List<IndexEntry> srcIndexEntries, int startIndex) {
+    long startOffset = srcIndexEntries.get(startIndex).getValue().getOffset().getOffset();
+    int endIndex;
+    // [startIndex, endIndex) for a bundle of read.
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: This comment is not very meaningful. Particularly, since elsewhere `endIndex` seems to be inclusive for bundled read.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 26]
```diff
@@ -733,53 +761,94 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+          bufferToUse.limit((int) (
+              srcIndexEntries.get(endIndex).getValue().getOffset().getOffset() + srcIndexEntries.get(endIndex)
+                  .getValue()
+                  .getSize() - startOffset));
+        }
+        // do IO read
+        int bytesRead = fileChannel.read(bufferToUse, startOffset);
+        if (bytesRead != bufferToUse.limit()) {
+          srcMetrics.compactionBundleReadBufferReadExtra.inc();
+          logger.warn("fileChannel is reading more than expected: {}/{}", bufferToUse.limit(), bytesRead);
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: On the contrary, reaching here means `fileChannel` read **less** than what `bufferToUse` is capable of holding, not more.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 27]
```diff
@@ -733,53 +761,94 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
+        }
+
+        // copy from buffer to tgtLog
+        for (int i = startIndex; i <= endIndex; i++) {
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: To summarize, there are 3 cases:
Case 1: `bytesRead > bufferToUse.limit()` is guaranteed to never happen.
Case 2: `bytesRead == bufferToUse.limit()` is likely the common case.
Case 3: `bytesRead < bufferToUse.limit()` is a real possibility as per `FileChannel::read(ByteBuffer)` contract, and I don't think you are handling that. I think you are assuming Case 2 for the following code, which is incorrect.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 28]
```diff
@@ -717,6 +723,28 @@ private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws Sto
...
 
+  /**
+   * Calculate the farthest index that can be used for a bundle IO read based on startIndex and bundleReadBuffer capacity.
+   * @param srcIndexEntries all available index entries.
+   * @param startIndex the startIndex for a bundle read.
+   * @return the endIndex which can be used for a bundle read.
+   */
+  private int getBundleReadEndIndex(List<IndexEntry> srcIndexEntries, int startIndex) {
+    long startOffset = srcIndexEntries.get(startIndex).getValue().getOffset().getOffset();
+    int endIndex;
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: I would use `i` and `j`, rather than `startIndex` and `endIndex`. Also `endIndex` is just an index and only becomes the "end" index at the end of the iteration. I think `j` is more appropriate.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 29]
```diff
@@ -733,53 +761,94 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
...
-          Offset endOffsetOfLastMessage = tgtLog.getEndOffset();
-          // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
-          diskIOScheduler.getSlice(DiskManager.CLEANUP_OPS_JOB_NAME, DiskManager.CLEANUP_OPS_JOB_NAME, writtenLastTime);
-          tgtLog.appendFrom(fileChannel, srcValue.getSize());
-          FileSpan fileSpan = tgtLog.getFileSpanForMessage(endOffsetOfLastMessage, srcValue.getSize());
-          if (srcValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
-            IndexValue putValue = tgtIndex.findKey(srcIndexEntry.getKey());
-            if (putValue != null) {
-              tgtIndex.markAsDeleted(srcIndexEntry.getKey(), fileSpan, srcValue.getOperationTimeInMs());
+      int startIndex = 0;
```
[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: How about `start`, `end` with some comment? 

`i, j` are meaningless, plus i is also used in the inner loop.

**Comment by zzmao** [comment]
@pnarayanan comments addressed. 
1. `start, end` are used instead of `i, j`
2. `ReadableChannel` can't seek. It only provides `read(dst)` method. 
A fileChannel version util function is added. 
`public static void readFileToByteBuffer(FileChannel fileChannel, long offset, ByteBuffer buffer)
  `

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 30]
```diff
@@ -717,6 +725,27 @@ private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws Sto
         IndexValue.Flags.Delete_Index));
   }
 
+  /**
+   * Calculate the farthest index that can be used for a bundle IO read based on start index and bundleReadBuffer capacity.
+   * @param sortedSrcIndexEntries all available entries, which are ordered by offset.
+   * @param start the staring index for the given list, not {@link IndexEntry}.
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: "starting"

Comment (pnarayanan) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: I think the "not IndexEntry" is unnecessary.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 31]
```diff
@@ -717,6 +725,27 @@ private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws Sto
...
+      long currentSize =
+          sortedSrcIndexEntries.get(end).getValue().getOffset().getOffset() + sortedSrcIndexEntries.get(end)
+              .getValue()
+              .getSize() - startOffset;
+      if (currentSize > bundleReadBuffer.capacity()) {
+        break;
+      }
+    }
+    return end - 1;
+  }
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: The java streams way of achieving this looks to be the following (just for information as some might find it more readable and clean - no need to make the change):
```java
 return IntStream.range(0, sortedSrcIndexEntries.size())
        .filter(i -> sortedSrcIndexEntries.get(i).getValue().getOffset().getOffset() + 
                     sortedSrcIndexEntries.get(i).getValue().getSize() - startOffset > 
                     bundleReadBuffer.capacity())
        .findFirst()
        .orElse(sortedSrcIndexEntries.size()) - 1;
```

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 32]
```diff
@@ -721,6 +721,27 @@ public static long addSecondsToEpochTime(long epochTimeInMs, long deltaTimeInSec
...
+   * @param offset starting offset of the fileChanel to be read from
+   * @param buffer the destination of the read
+   * @throws IOException
+   */
+  public static void readFileToByteBuffer(FileChannel fileChannel, long offset, ByteBuffer buffer) throws IOException {
+    int read = 0;
+    int expectedRead = buffer.remaining();
+    while (buffer.hasRemaining()) {
+      int sizeRead = fileChannel.read(buffer, offset);
+      if (sizeRead == 0 || sizeRead == -1) {
```
Comment (pnarayanan) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: Is `size == 0` is an exception scenario? (as in, is `FileChannel` necessarily blocking)?

Comment (zzmao) [reply ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: I thought fileChannel.read() return 0 only when offset is valid offset and buffer has no space. (This should never happen because we have the check `buffer.hasRemaining()`) if this happen, that's something unexpected, we should throw exception. 

But you are making a good point, if FileChannel is non-blocking, it may return 0. Then I got this: 
https://stackoverflow.com/questions/3955250/why-filechannel-in-java-is-not-non-blocking

Looks like fileChannel is necessarily blocking.

Comment (pnarayanan) [reply ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: Could you log sizeRead as well in the log message?

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 33]
```diff
@@ -721,6 +721,27 @@ public static long addSecondsToEpochTime(long epochTimeInMs, long deltaTimeInSec
...
 
+  /**
+   * Read {@link FileChannel} from a given offset to fill up a {@link ByteBuffer}. Total bytes read should be buffer's
+   * remaining size(limit - position).
+   * @param fileChannel from which data to be read from
+   * @param offset starting offset of the fileChanel to be read from
+   * @param buffer the destination of the read
+   * @throws IOException
+   */
+  public static void readFileToByteBuffer(FileChannel fileChannel, long offset, ByteBuffer buffer) throws IOException {
```
Comment (vgkholla) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: add a test for this (if not already there)
